README


This archive contains the distribution IO-Storm,
version 0.15:

  IO::Storm allows you to write Bolts and Spouts for Storm in Perl.

This software is copyright (c) 2014 by Educational Testing Service.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.029.



CHANGES
==================================================
Changes from 2013-12-16 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 0.15 at 2014-12-16 02:59:36 +0000
-----------------------------------------

  Change: 6c58d4674f7a92e472a0b1a61f5615ebd06c5dd9
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-12-15 14:50:03 +0000

    Bump version to 0.15 

  Change: 8fd8fd0e6e28d587ac679f48d52a4e1604065978
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-12-15 14:49:55 +0000

    Switch to binmode from open pragma. Should fix UTF-8 issues. 

-----------------------------------------
version 0.14 at 2014-12-12 15:46:56 +0000
-----------------------------------------

  Change: e53022157e7fc92b2f38e12d3328d8afd804110c
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-12-12 10:46:56 +0000

    Bump version to 0.14 

  Change: ea61f60ca73b58fa5a2661089c6099993e23241e
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-12-12 10:45:24 +0000

    IO::Storm now uses UTF-8 encoding for all communication with Storm.

    It should have been this way all along. 

  Change: 2cd2934bd56842635308d57fcbe78a4c46505f55
  Author: Dan Blanchard <dan.blanchard@gmail.com>
  Date : 2014-12-04 11:55:53 +0000

    Merge pull request #2 from dapuck/master

    Auto anchor inserts entire tuples instead of just the ids 

  Change: 7d49adace5788faea700f3de6d444b28b16b509d
  Author: Ian McCall <imccall@where2getit.com>
  Date : 2014-12-04 09:58:55 +0000

    Merge remote-tracking branch 'upstream/master'

    Merging in travis file from parent repo. 

  Change: e6eff8ee9f73d4e43e2f7b0fdf5fb5c814cbee3d
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-12-04 10:35:39 +0000

    Add .travis.yml for automated CI 

  Change: 0c994e40bdb28eb528742799339b19589ab03519
  Author: Ian McCall <imccall@where2getit.com>
  Date : 2014-12-03 16:02:51 +0000

    Fixed use of tab instead of spaces. 

  Change: 1cda57e8ae4472e7318cbd1b8a794dac57d78d3f
  Author: Ian McCall <imccall@where2getit.com>
  Date : 2014-12-03 15:49:22 +0000

    Fixes auto anchoring. Refactor emit.

    When using auto_anchor the entire tuple gets added to the anchor
    array instead of just the id. This change fixes that, and makes it so
    that the anchors arg will override auto_anchor. Also refactored the
    code to reflect that args will be defined, even if the user doesn't
    supply it. 

  Change: 2d5fe541d99d98e87d2b8e2f8442ba5d416bfa06
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-08-12 22:27:27 +0000

    Merge branch 'master' of github.com:dan-blanchard/io-storm 

  Change: c06fb9790ff218bc950122719c710ea49df653a7
  Author: Dan Blanchard <dan.blanchard@gmail.com>
  Date : 2014-08-12 13:51:08 +0000

    Fix typo in README from running markdown2pod 

-----------------------------------------
version 0.13 at 2014-08-13 02:23:24 +0000
-----------------------------------------

  Change: c5964b0ba703ed5c0a6f1e9909b2ade2b03884d3
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-08-12 22:23:24 +0000

    Update README 

  Change: ad60c74e7237c36dddd2b43c1d37c03a399d449a
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-08-12 22:22:47 +0000

    Fix yet another issue caused by operations on array refs being marked
    experimental in 5.20 

  Change: 823147757180927f18f607ce463cc6210d4788e8
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-08-12 10:46:53 +0000

    Fix a typo in POD for Storm.pm. 

-----------------------------------------
version 0.12 at 2014-08-12 14:41:38 +0000
-----------------------------------------

  Change: 1c5150f46c5abd0a4f0325165d3d611c6f3b2484
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-08-12 10:41:38 +0000

    Add .build to .gitignore 

  Change: 6babb9d271a32309271d82908320dd7fae9f1155
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-08-12 10:40:52 +0000

    Change minimum Perl version back to 5.10 and fix compatibility. 

-----------------------------------------
version 0.11 at 2014-08-11 15:49:50 +0000
-----------------------------------------

  Change: 47bcb5993015c74e316cb045c55380f5770611f0
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-08-11 11:49:50 +0000

    Bump required Perl version up to 5.14 

-----------------------------------------
version 0.10 at 2014-08-11 14:34:10 +0000
-----------------------------------------

  Change: 1432cd7bae328b9c6bbc5096ee505969e05643b7
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-08-11 10:34:10 +0000

    Make explicit Component package and add POD documentation to Storm.pm 

  Change: 2d24e984ee857024a3de616923fddba13fc4e563
  Author: Dan Blanchard <dan.blanchard@gmail.com>
  Date : 2014-08-09 07:56:53 +0000

    Update dist.ini

    Added autoprereqs and GitHub-specific plugins for managing README and
    issues. 

-----------------------------------------
version 0.09 at 2014-08-09 03:19:54 +0000
-----------------------------------------

  Change: 4b21e54221621d66b83531d187c1960042fed227
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-08-08 23:19:54 +0000

    v0.09 

-----------------------------------------
version 0.08 at 2014-08-09 03:15:33 +0000
-----------------------------------------

  Change: 9cfc8530f3864455a9e66e155d46da69b15988a6
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-08-08 23:15:33 +0000

    Remove version numbers from Bolt.pm, Spout.pm, and Tuple.pm 

-----------------------------------------
version 0.07 at 2014-08-09 03:09:45 +0000
-----------------------------------------

  Change: 5d83b0fb597139e4768efbc6d12e766bd932948a
  Author: Daniel Blanchard <dblanchard@ets.org>
  Date : 2014-08-08 23:09:45 +0000

    Tweak a bunch of Dist::Zilla settings. 

  Change: ebc79f18c1ec263b59b2d3910fbedbd23d884c6e
  Author: Daniel Blanchard <dblanchard@ets.org>

Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.029.
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "IO::Storm allows you to write Bolts and Spouts for Storm in Perl.",
  "AUTHOR" => "Dan Blanchard <dblanchard\@ets.org>, Cory G Watson <gphat\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "IO-Storm",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010",
  "NAME" => "IO::Storm",
  "PREREQ_PM" => {
    "Data::Dumper" => 0,
    "IO::File" => 0,
    "IO::Handle" => 0,
    "JSON::XS" => 0,
    "Log::Log4perl" => 0,
    "Moo" => 0,
    "Try::Tiny" => 0,
    "namespace::clean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Test::MockObject" => 0,
    "Test::More" => 0,
    "Test::Output" => 0
  },
  "VERSION" => "0.15",
  "test" => {
    "TESTS" => "t/IO/Storm/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Dumper" => 0,
  "ExtUtils::MakeMaker" => 0,
  "IO::File" => 0,
  "IO::Handle" => 0,
  "JSON::XS" => 0,
  "Log::Log4perl" => 0,
  "Moo" => 0,
  "Test::MockObject" => 0,
  "Test::More" => 0,
  "Test::Output" => 0,
  "Try::Tiny" => 0,
  "namespace::clean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/IO/Storm.pm
use v5.10;
use strict;
use warnings;

package IO::Storm;
# ABSTRACT: IO::Storm allows you to write Bolts and Spouts for Storm in Perl.
$IO::Storm::VERSION = '0.15';
1;

__END__

=pod

=head1 NAME

IO::Storm - IO::Storm allows you to write Bolts and Spouts for Storm in Perl.

=head1 VERSION

version 0.15

=head1 SYNOPSIS

    package SplitSentenceBolt;

    use Moo;
    use namespace::clean;

    extends 'Storm::Bolt';

    sub process {
        my ($self, $tuple) = @_;

        my @words = split(' ', $tuple->values->[0]);
        foreach my $word (@words) {

            $self->emit([ $word ]);
        }

    }

    SplitSentenceBolt->new->run;

=head1 DESCRIPTION

IO::Storm allows you to leverage Storm's multilang support to write Bolts and
Spouts in Perl.  As of version 0.02, the API is designed to very closely mirror
that of the L<Streamparse Python library|http://streamparse.readthedocs.org/en/latest/api.html>.  The exception being that we don't currently support
the C<BatchingBolt> class or the C<emit_many> methods.

=encoding utf8

=head1 Bolts

To create a Bolt, you want to extend the C<Storm::Bolt> class.

    package SplitSentenceBolt;

    use Moo;
    use namespace::clean;

    extends 'Storm::Bolt';

=head2 Processing tuples

To have your Bolt start processing tuples, you want to override the C<process>
method, which takes a C<IO::Storm::Tuple> as its only argument.  This method
should do any processing you want to perform on the tuple and then C<emit> its
output.

    sub process {
        my ($self, $tuple) = @_;

        my @words = split(' ', $tuple->values->[0]);
        foreach my $word (@words) {

            $self->emit([ $word ]);
        }

    }

To actually start your Bolt, call the C<run> method, which will initialize the
bolt and start the event loop.

    SplitSentenceBolt->new->run;

=head2 Automatic reliability

By default, the Bolt will automatically handle acks, anchoring, and
failures.  If you would like to customize the behavior of any of these things,
you will need to set the C<auto_anchor>, C<auto_anchor>, or C<auto_fail>
attributes to 0.  For more information about Storm's guaranteed message
processing, please L<see their documentation|https://storm.incubator.apache.org/documentation/Guaranteeing-message-processing.html#what-is-storms-reliability-api>.

=head1 Spouts

To create a Spout, you want to extend the C<Storm::Spout> class.

    package SentenceSpout;

    use Moo;
    use namespace::clean;

    extends 'Storm::Spout';

=head2 Emitting tuples

To actually emit anything on your Spout, you have to implement the
C<next_tuple> method.

    my $sentences = ["a little brown dog",
                     "the man petted the dog",
                     "four score and seven years ago",
                     "an apple a day keeps the doctor away",];
    my $num_sentences = scalar(@$sentences);

    sub next_tuple {
        my ($self) = @_;

        $self->emit( [ $sentences->[ rand($num_sentences) ] ] );

    }

To actually start your Spout, call the C<run> method, which will initialize the
Spout and start the event loop.

    SentenceSpout->new->run;

=head1 Methods supported by Spouts and Bolts

=head2 Custom initialization

If you need to have some custom action happen when your component is being
initialized, just override C<initialize> method, which receives the Storm
configuration for the component and information about its place in the topology
as its arguments.

    sub initialize {
        my ( $self, $storm_conf, $context ) = @_;
    }

=head2 Logging

Use the C<log> method to send messages back to the Storm ShellBolt parent
process which will be added to the general Storm log.

    sub process {
        my ($self, $tuple) = @_;
        ...
        $self->log("Working on $tuple");
        ...
    }

=head1 AUTHORS

=over 4

=item *

Dan Blanchard <dblanchard@ets.org>

=item *

Cory G Watson <gphat@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Educational Testing Service.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/IO/Storm/Spout.pm
# ABSTRACT: The base class for all IO::Storm Spout.

package IO::Storm::Spout;
$IO::Storm::Spout::VERSION = '0.15';
# Imports
use strict;
use warnings;
use v5.10;

use Moo;
use namespace::clean;

extends 'IO::Storm::Component';


sub initialize {
    my ( $self, $storm_conf, $context ) = @_;
}


sub ack {
    my ( $self, $id ) = @_;
}


sub fail {
    my ( $self, $id ) = @_;
}


sub next_tuple {
    my ($self) = @_;
}


sub emit ($$;$) {
    my ( $self, $tuple, $args ) = @_;

    my $msg = { command => 'emit', tuple => $tuple };

    if ( defined($args) ) {
        if ( defined( $args->{tup_id} ) ) {
            $msg->{id} = $args->{tup_id};
        }
        if ( defined( $args->{stream} ) ) {
            $msg->{stream} = $args->{stream};
        }
        if ( defined( $args->{direct_task} ) ) {
            $msg->{task} = $args->{direct_task};
        }
    }

    $self->send_message($msg);

    if ( defined $msg->{task} ) {
        return $msg->{task};
    }
    else {
        return $self->read_task_ids();
    }
}


sub run {
    my ($self) = @_;

    my ( $storm_conf, $context ) = $self->read_handshake();
    $self->_setup_component( $storm_conf, $context );
    $self->initialize( $storm_conf, $context );

    while (1) {
        my $msg = $self->read_command();
        if ( $msg->{command} eq 'next' ) {
            $self->next_tuple;
        }
        elsif ( $msg->{command} eq 'ack' ) {
            $self->ack( $msg->{id} );
        }
        elsif ( $msg->{command} eq 'fail' ) {
            $self->fail( $msg->{id} );
        }
        $self->sync();
    }
}

1;

__END__

=pod

=head1 NAME

IO::Storm::Spout - The base class for all IO::Storm Spout.

=head1 VERSION

version 0.15

=head1 NAME

IO::Storm::Spout - The base class for all IO::Storm Spout.

=head1 VERSION

version 0.06

=head1 METHODS

=head2 initialize

Called immediately after the initial handshake with Storm and before the main
run loop. A good place to initialize connections to data sources.

=head2 ack

Called when a bolt acknowledges a tuple in the topology.

=head2 fail

Called when a tuple fails in the topology

A Spout can choose to emit the tuple again or ignore the fail. The default is
to ignore.

=head2 next_tuple

Implement this function to emit tuples as necessary.

This function should not block, or Storm will think the spout is dead. Instead,
let it return and streamparse will send a noop to storm, which lets it know the
spout is functioning.

=head2 emit

Emit a spout tuple message.

:param tup: the tuple to send to Storm.  Should contain only
            JSON-serializable data.
:type tup: list
:param tup_id: the ID for the tuple. Leave this blank for an
               unreliable emit.
:type tup_id: str
:param stream: ID of the stream this tuple should be emitted to.
               Leave empty to emit to the default stream.
:type stream: str
:param direct_task: the task to send the tuple to if performing a
                    direct emit.
:type direct_task: int

=head2 run

Main run loop for all spouts.

Performs initial handshake with Storm and reads tuples handing them off to
subclasses.  Any exceptions are caught and logged back to Storm prior to the
Perl process exiting.

Subclasses should **not** override this method.

=head1 AUTHORS

=over 4

=item *

Cory G Watson <gphat@cpan.org>

=item *

Dan Blanchard <dblanchard@ets.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Infinity Interactive, Inc.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 AUTHORS

=over 4

=item *

Dan Blanchard <dblanchard@ets.org>

=item *

Cory G Watson <gphat@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Educational Testing Service.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/IO/Storm/Tuple.pm
# ABSTRACT: Storm's primitive data type passed around via streams.

package IO::Storm::Tuple;
$IO::Storm::Tuple::VERSION = '0.15';
# Imports
use strict;
use warnings;
use v5.10;

# Setup Moo for object-oriented niceties
use Moo;
use namespace::clean;

has 'id' => ( is => 'rw' );

has 'component' => ( is => 'rw' );

has 'stream' => ( is => 'rw' );

has 'task' => ( is => 'rw' );

has 'values' => ( is => 'rw' );

sub TO_JSON {
    my ($self) = @_;
    return {
        id        => $self->id,
        component => $self->component,
        stream    => $self->stream,
        task      => $self->task,
        values    => $self->values
    };
}

1;

__END__

=pod

=head1 NAME

IO::Storm::Tuple - Storm's primitive data type passed around via streams.

=head1 VERSION

version 0.15

=head1 NAME

IO::Storm::Tuple - Storm's primitive data type passed around via streams.

=head1 VERSION

version 0.06

=head1 AUTHORS

=over 4

=item *

Cory G Watson <gphat@cpan.org>

=item *

Dan Blanchard <dblanchard@ets.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Infinity Interactive, Inc.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 AUTHORS

=over 4

=item *

Dan Blanchard <dblanchard@ets.org>

=item *

Cory G Watson <gphat@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Educational Testing Service.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/IO/Storm/BasicBolt.pm
package IO::Storm::BasicBolt;
{
  $IO::Storm::BasicBolt::VERSION = '0.01';
}
use Moose;

extends 'IO::Storm';

use Log::Log4perl;

my $logger = Log::Log4perl->get_logger('storm.basicbolt');

sub process {
    my ($self, $tuple) = @_;
}
    
sub run {
    my ($self) = @_;
    # XXX

    my ($conf, $context) = $self->init_bolt;
    while(1) {
        my $tup = $self->read_tuple;
        $self->_anchor($tup);
        $self->process($tup);
        $self->ack($tup);
        $self->sync;
    }
}

1;
__END__
=pod

=head1 NAME

IO::Storm::BasicBolt

=head1 VERSION

version 0.01

=head1 AUTHOR

Cory G Watson <gphat@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Infinity Interactive, Inc.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/IO/Storm/Component.pm
# ABSTRACT: The base class for Bolts and Spouts.

package IO::Storm::Component;
$IO::Storm::Component::VERSION = '0.15';
# Imports
use strict;
use warnings;
use v5.10;
use IO::Handle qw(autoflush);
use IO::File;
use Log::Log4perl qw(:easy);
use JSON::XS;
use Data::Dumper;
use IO::Storm::Tuple;

# Setup Moo for object-oriented niceties
use Moo;
use namespace::clean;

# Setup STDIN/STDOUT/STDERR to use UTF8
binmode STDERR, ':utf8';
binmode STDIN,  ':encoding(UTF-8)';
binmode STDOUT, ':utf8';

has '_pending_commands' => (
    is      => 'rw',
    default => sub { [] },
);

has '_pending_taskids' => (
    is      => 'rw',
    default => sub { [] },
);

has '_stdin' => (
    is      => 'rw',
    default => sub {
        my $io = IO::Handle->new;
        $io->fdopen( fileno(STDIN), 'r' );
    }
);

has 'max_lines' => (
    is      => 'rw',
    default => 100
);

has 'max_blank_msgs' => (
    is      => 'rw',
    default => 500
);

has '_json' => (
    is      => 'rw',
    default => sub { JSON::XS->new->allow_blessed->convert_blessed }
);

has '_topology_name' => (
    is        => 'rw',
    init_args => undef
);

has '_task_id' => (
    is        => 'rw',
    init_args => undef
);

has '_component_name' => (
    is        => 'rw',
    init_args => undef
);

has '_debug' => (
    is        => 'rw',
    init_args => undef
);

has '_storm_conf' => (
    is        => 'rw',
    init_args => undef
);

has '_context' => (
    is        => 'rw',
    init_args => undef
);

my $logger = Log::Log4perl->get_logger('storm');


sub _setup_component {
    my ( $self, $storm_conf, $context ) = @_;
    my $conf_is_hash = ref($storm_conf) eq ref {};
    $self->_topology_name(
        ( $conf_is_hash && exists( $storm_conf->{'topology.name'} ) )
        ? $storm_conf->{'topology.name'}
        : ''
    );
    $self->_task_id( exists( $context->{taskid} ) ? $context->{taskid} : '' );
    $self->_component_name('');
    if ( exists( $context->{'task->component'} ) ) {
        my $task_component = $context->{'task->component'};
        if ( exists( $task_component->{ $self->_task_id } ) ) {
            $self->_component_name( $task_component->{ $self->_task_id } );
        }
    }
    $self->_debug(
        ( $conf_is_hash && exists( $storm_conf->{'topology.debug'} ) )
        ? $storm_conf->{'topology.debug'}
        : 0
    );
    $self->_storm_conf($storm_conf);
    $self->_context($context);
}


sub read_message {
    $logger->debug('start read_message');
    my $self         = shift;
    my $blank_lines  = 0;
    my $message_size = 0;
    my $line         = '';

    my @messages = ();
    while (1) {
        $line = $self->_stdin->getline;
        if ( defined($line) ) {
            $logger->debug("read_message: line=$line");
        }
        else {
            $logger->error( "Received EOF while trying to read stdin from "
                    . "Storm, pipe appears to be broken, exiting." );
            exit(1);
        }
        if ( $line eq "end\n" ) {
            last;
        }
        elsif ( $line eq '' ) {
            $logger->error( "Received EOF while trying to read stdin from "
                    . "Storm, pipe appears to be broken, exiting." );
            exit(1);
        }
        elsif ( $line eq "\n" ) {
            $blank_lines++;
            if ( $blank_lines % 1000 == 0 ) {
                $logger->warn( "While trying to read a command or pending "
                        . "task ID, Storm has instead sent $blank_lines "
                        . "'\\n' messages." );
                next;
            }
        }
        chomp($line);
        push( @messages, $line );
    }

    return $self->_json->decode( join( "\n", @messages ) );
}

sub read_task_ids {
    my $self = shift;

    if ( scalar( @{ $self->_pending_taskids } ) ) {
        return shift( @{ $self->_pending_taskids } );
    }
    else {
        my $msg = $self->read_message();
        while ( ref($msg) ne 'ARRAY' ) {
            push( @{ $self->_pending_commands }, $msg );
            $msg = $self->read_message();
        }

        return $msg;
    }
}

sub read_command {
    my $self = shift;

    if ( @{ $self->_pending_commands } ) {
        return shift( @{ $self->_pending_commands } );
    }
    else {
        my $msg = $self->read_message();
        while ( ref($msg) eq 'ARRAY' ) {
            push( @{ $self->_pending_taskids }, $msg );
            $msg = $self->read_message();
        }
        return $msg;
    }
}


sub read_tuple {
    my $self = shift;
    $logger->debug('read_tuple');

    my $tupmap = $self->read_command();

    return IO::Storm::Tuple->new(
        id        => $tupmap->{id},
        component => $tupmap->{comp},
        stream    => $tupmap->{stream},
        task      => $tupmap->{task},
        values    => $tupmap->{tuple}
    );
}


sub read_handshake {
    my $self = shift;

    # TODO: Figure out how to redirect stdout to ensure that print
    # statements/functions won't crash the Storm Java worker

    autoflush STDOUT 1;
    autoflush STDERR 1;

    my $msg = $self->read_message();
    $logger->debug(
        sub { 'Received initial handshake from Storm: ' . Dumper($msg) } );

    # Write a blank PID file out to the pidDir
    my $pid      = $$;
    my $pid_dir  = $msg->{pidDir};
    my $filename = $pid_dir . '/' . $pid;
    open my $fh, '>', $filename
        or die "Cant't write to '$filename': $!\n";
    $fh->close;
    $logger->debug("Sending process ID $pid to Storm");
    $self->send_message( { pid => int($pid) } );

    return [ $msg->{conf}, $msg->{context} ];
}


sub send_message {
    my ( $self, $msg ) = @_;
    say $self->_json->encode($msg);
    say "end";
}


sub sync {
    my $self = shift;
    $self->send_message( { command => 'sync' } );
}


sub log {
    my ( $self, $message ) = @_;
    $self->send_message( { command => 'log', msg => $message } );
}

1;

__END__

=pod

=head1 NAME

IO::Storm::Component - The base class for Bolts and Spouts.

=head1 VERSION

version 0.15

=head1 METHODS

=head2 _setup_component

Add helpful instance variables to component after initial handshake with Storm.

=head2 read_message

Read a message from the ShellBolt.  Reads until it finds a "end" line.

=head2 read_tuple

Turn the incoming Tuple structure into an <IO::Storm::Tuple>.

=head2 read_handshake

Read and process an initial handshake message from Storm

=head2 send_message

Send a message to Storm, encoding it as JSON.

=head2 sync

Send a sync command to Storm.

=head2 log

Send a log command to Storm

=head1 AUTHORS

=over 4

=item *

Dan Blanchard <dblanchard@ets.org>

=item *

Cory G Watson <gphat@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Educational Testing Service.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/IO/Storm/Bolt.pm
# ABSTRACT: The base class for all IO::Storm Bolts.

package IO::Storm::Bolt;
$IO::Storm::Bolt::VERSION = '0.15';
# Imports
use strict;
use warnings;
use v5.10;
use Try::Tiny;

# Setup Moo for object-oriented niceties
use Moo;
use namespace::clean;


extends 'IO::Storm::Component';

# A boolean indicating whether or not the bolt should automatically
# anchor emits to the incoming tuple ID. Tuple anchoring is how Storm
# provides reliability, you can read more about tuple anchoring in Storm's
# docs:
# https://storm.incubator.apache.org/documentation/Guaranteeing-message-processing.html#what-is-storms-reliability-api
has 'auto_anchor' => (
    is      => 'rw',
    default => 1
);

# A boolean indicating whether or not the bolt should automatically
# acknowledge tuples after ``process()`` is called.
has 'auto_ack' => (
    is      => 'rw',
    default => 1
);

# A boolean indicating whether or not the bolt should automatically fail
# tuples when an exception occurs when the ``process()`` method is called.
has 'auto_fail' => (
    is      => 'rw',
    default => 1
);

# Using a list so Bolt and subclasses can have more than one current_tup
has '_current_tups' => (
    is       => 'rw',
    default  => sub { [] },
    init_arg => undef
);


sub initialize {
    my ( $self, $storm_conf, $context ) = @_;
}


sub process {
    my ( $self, $tuple ) = @_;
}


sub emit ($$;$) {
    my ( $self, $tuple, $args ) = @_;

    $args = $args // {};
    my $msg = { command => 'emit', tuple => $tuple };

    my $anchors = [];
    if ( $self->auto_anchor ) {
        $anchors = $self->_current_tups // [];
    }
    unless ( defined( $args->{anchors} ) ) {
    	$args->{anchors} = $anchors;
    }

    my $a;
    for $a ( @{ $args->{anchors} } ) {
        if ( ref($a) eq "IO::Storm::Tuple" ) {
            $a = $a->id;
        }
        push( @$anchors, $a );
    }

    if ( defined( $args->{stream} ) ) {
        $msg->{stream} = $args->{stream};
    }

    if ( defined( $args->{direct_task} ) ) {
        $msg->{task} = $args->{direct_task};
    }

    $msg->{anchors} = $anchors;

    $self->send_message($msg);

    if ( defined $msg->{task} ) {
        return $msg->{task};
    }
    else {
        return $self->read_task_ids();
    }
}


sub ack {
    my ( $self, $tuple ) = @_;
    my $tup_id;
    if ( ref($tuple) eq "IO::Storm::Tuple" ) {
        $tup_id = $tuple->id;
    }
    else {
        $tup_id = $tuple;
    }
    $self->send_message( { command => 'ack', id => $tup_id } );
}


sub fail {
    my ( $self, $tuple ) = @_;
    my $tup_id;
    if ( ref($tuple) eq "IO::Storm::Tuple" ) {
        $tup_id = $tuple->id;
    }
    else {
        $tup_id = $tuple;
    }

    $self->send_message( { command => 'fail', id => $tup_id } );
}


sub run {
    my ($self) = @_;
    my $tup;

    my ( $storm_conf, $context ) = $self->read_handshake();
    $self->_setup_component( $storm_conf, $context );
    $self->initialize( $storm_conf, $context );

    try {
        while (1) {
            $tup = $self->read_tuple();
            $self->_current_tups( [$tup] );
            $self->process($tup);
            if ( $self->auto_ack ) {
                $self->ack($tup);
            }

            # reset so that we don't accidentally fail the wrong tuples
            # if a successive call to read_tuple fails
            $self->_current_tups( [] );
        }
    }
    catch {
        my $error = $_;
        if ( scalar( @{ $self->_current_tups } ) == 1 ) {
            $tup = $self->_current_tups->[0];
            if ( $self->auto_fail ) {
                $self->fail($tup);
            }
        }
        $self->log("Bolt encountered exception: $_");
        die("Encounter exception in Bolt: $_");
    };
}

1;

__END__

=pod

=head1 NAME

IO::Storm::Bolt - The base class for all IO::Storm Bolts.

=head1 VERSION

version 0.15

=head1 NAME

IO::Storm::Bolt - The base class for all IO::Storm Bolts.

=head1 VERSION

version 0.06

=head1 METHODS

=head2 initialize

Called immediately after the initial handshake with Storm and before the main
run loop. A good place to initialize connections to data sources.

=head2 process

Process a single tuple of input. This should be overriden by subclasses

=head2 emit

Emit a tuple to a stream.

:param tuple: the Tuple payload to send to Storm, should contain only
            JSON-serializable data.
:type tuple: arrayref
:param stream: the ID of the stream to emit this tuple to. Specify
               ``undef`` to emit to default stream.
:type stream: scalar
:param anchors: IDs of the tuples (or the <IO::Storm::Tuple> instances) which
                the emitted tuples should be anchored to. If ``auto_anchor`` is
                set and you have not specified ``anchors``, ``anchors`` will be
                set to the incoming/most recent tuple ID(s).
:type anchors: arrayref
:param direct_task: the task to send the tuple to.
:type direct_task: scalar

=head2 ack

Acknowledge a tuple. Argument can be either a Tuple or an ID.

=head2 fail

Fail a tuple. Argument can be either a Tuple or an ID.

=head2 run

Main run loop for all bolts.

Performs initial handshake with Storm and reads tuples handing them off
to subclasses.  Any exceptions are caught and logged back to Storm
prior to the Perl process exiting.

Subclasses should **not** override this method.

=head1 AUTHORS

=over 4

=item *

Cory G Watson <gphat@cpan.org>

=item *

Dan Blanchard <dblanchard@ets.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Infinity Interactive, Inc.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 AUTHORS

=over 4

=item *

Dan Blanchard <dblanchard@ets.org>

=item *

Cory G Watson <gphat@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Educational Testing Service.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
